### ISSUE TITLE
Support negative icon-padding/text-padding values

### ISSUE THREAD

**Issue by samanpwbb** [mapbox/mapbox-gl-js#6370]
**mapbox-gl-js version**: v0.44.1

### Steps to Trigger Behavior

Set `icon-padding` to -2 on a style layer.

### Expected Behavior

Calculate an allowed overlap between icons of -2px. Red "X"'s are icons that are blocked by the padding rule while black "X"'s are rendered:

![screen shot 2018-03-20 at 6 05 54 pm](https://user-images.githubusercontent.com/108094/37685338-7b9d85a0-2c69-11e8-8b10-a1407354383a.png)

This would be useful to create higher density maps where a little overlap helps in order to express that density, but a lot of overlap would look sloppy and unintentional.

### Actual Behavior

See "-2 is less than the minimum value 0" error.

**Comment by nickidlugash** [comment]
@ansis Any thoughts on whether it would make sense to allow negative padding values? I think this might be useful for cases where you want overlapping symbols that have both an icon and text (e.g. stacked markers/pins). I think the general expectation in these cases might be that you could set the icon and text to have the same collision value (`-allow-overlap` or `-ignore-placement`) and that the symbols would visually overlap as if the text and icon were one entity.  Since that kind of drawing behavior is not practical to implement, however, perhaps negative padding in conjunction with allowing collision detection for text might achieve a similar result. I discussed this briefly with @ChrisLoer, who vaguely recalled discussing this with you at one point.

**Comment by stdmn** [comment]
+1 on this request. You can see from the screenshot that there are a couple of conditions where I would expect there to be at least one symbol left after collision detection. Having a negative padding option would offer a little bit more control over situations where it's OK if there is a little bit of overlap

<img width="960" alt="screen shot 2018-12-15 at 5 36 31 pm" src="https://user-images.githubusercontent.com/17128916/50048061-9ebd2f80-0090-11e9-8c73-4201c52edc8d.png">
<img width="960" alt="screen shot 2018-12-15 at 5 22 07 pm" src="https://user-images.githubusercontent.com/17128916/50048062-9ebd2f80-0090-11e9-80dd-a341d7161c2d.png">

**Comment by stevage** [comment]
Huh. I never noticed the `icon-padding` attribute before. To be honest, I can't really imagine a use case for positive values - but I would love the option to set negative values. That's basically what I was advocating for in #8340.

**Comment by stevage** [comment]
Another real world example. It's really frustrating that my only two options are these two. What we really want is something in between the two.

![Screen Shot 2019-07-02 at 10 37 01 pm](https://user-images.githubusercontent.com/678344/60513226-02036c00-9d1a-11e9-99a7-52616bfc2d5c.png)

![Screen Shot 2019-07-02 at 10 37 09 pm](https://user-images.githubusercontent.com/678344/60513239-0af43d80-9d1a-11e9-86f7-45880f90cee1.png)

**Comment by SeasonedDaily** [comment]
> Another real world example. It's really frustrating that my only two options are these two. What we really want is something in between the two.
> 
> ![Screen Shot 2019-07-02 at 10 37 01 pm](https://user-images.githubusercontent.com/678344/60513226-02036c00-9d1a-11e9-99a7-52616bfc2d5c.png)
> 
> ![Screen Shot 2019-07-02 at 10 37 09 pm](https://user-images.githubusercontent.com/678344/60513239-0af43d80-9d1a-11e9-86f7-45880f90cee1.png)



> Another real world example. It's really frustrating that my only two options are these two. What we really want is something in between the two.
> 
> ![Screen Shot 2019-07-02 at 10 37 01 pm](https://user-images.githubusercontent.com/678344/60513226-02036c00-9d1a-11e9-99a7-52616bfc2d5c.png)
> 
> ![Screen Shot 2019-07-02 at 10 37 09 pm](https://user-images.githubusercontent.com/678344/60513239-0af43d80-9d1a-11e9-86f7-45880f90cee1.png)

You mentioned this is a real world example and I'm curious about where this is from? How did you end up resolving this?

**Comment by andrewharvey** [comment]
To add a use case, where the icon image contains a large shadow, overlapping of the shadow would be okay, but not the actual icon part. A negative `icon-padding` would solve it.
![2021-11-03_16-58](https://user-images.githubusercontent.com/117278/140015510-e1795250-3c8f-4d18-9da6-807439aa0f79.png)

**Comment by stevage** [comment]
@SeasonedDaily 
>You mentioned this is a real world example and I'm curious about where this is from? How did you end up resolving this?

It's from an internal site. We went with the non-overlapping version.

I still run into this all the time - very frustrating that you can't get icons to pack any tighter without having them overlap. 

Andrew's case is another good one. Although even there when you say "not the actual icon part", currently two circles won't pack very nicely corner to corner even though there is plenty of space there, because the hit-box testing is rectangular.

**Comment by SeasonedDaily** [comment]
We didn’t resolve. Developer dropped out and we switched solutions dropping
the feature. Not because of this though. Good luck!

**Comment by DavidMulder0** [comment]
I unintentionally posted a duplicate of this, but - as I mentioned in the duplicate - you can solve this using point clustering. It's a bit more work, but it does give you a lot of control. As much as I 100% agree that negative icon padding should be enabled, point clustering is really the 'correct' solution. To be fair, point clustering can be somewhat janky, but it does give you a lot of control over how you want to render overlapping markers.
